eq_cir_converter_service/utils/helper_utils.py

```diff
@@ -0,0 +1,181 @@
+        if len(paragraphs) == 1:
+            return paragraphs[0]  # return string if only one paragraph
+        return paragraphs
+    return clean_html_tags(text).strip()
+
+
+def process_text_object(
+    obj: dict[str, str | list | object],
+) -> dict[str, str | list | object] | list[str | dict[str, str | list | object]]:
+    """Processes a text object, cleaning HTML tags and extracting paragraphs with placeholders.
+
+    :param obj: A dictionary containing 'text' and possibly 'placeholders'.
+    :return: A cleaned text object or a list of paragraphs with placeholders.
+    """
+    if REGEX_PARA_SPLIT.search(str(obj.get("text", ""))):
+        # If the text contains <p> tags, split into paragraphs
+        # and clean each paragraph, extracting placeholders
+        return split_text_with_placeholders(obj)
+    obj["text"] = clean_html_tags(str(obj["text"])).strip()
+    return obj
+
+
+def process_list(elements: Sequence[str | Sequence[Any] | dict]) -> list[str | list | object]:
+    """Processes a list of elements, cleaning HTML tags and extracting paragraphs from text objects.
+
+    :param elements: A sequence of strings, lists, or dictionaries to process.
+    :return: A list of processed elements, which may include cleaned strings or lists of paragraphs.
+    """
+    result: list[str | list | object] = []
+
+    for item in elements:
+        if isinstance(item, dict):
+            expandable_key = None
+            for k, v in item.items():
+                if isinstance(v, str) and REGEX_PARA_SPLIT.search(v):
+                    # If the value is a string with <p> tags, extract paragraphs
+                    # and clean each paragraph
+                    # and extract placeholders
+                    expandable_key = k
+                    break
+
+            if expandable_key:
+                paragraphs = extract_paragraphs(item[expandable_key])
+                cleaned_paragraphs = [clean_html_tags(p).strip() for p in paragraphs if clean_html_tags(p).strip()]
+                result.extend({expandable_key: p} for p in cleaned_paragraphs)
+            else:
+                result.append(process_element(item))
+        else:
+            processed = process_element(item)
+            if isinstance(processed, list):
+                result.extend(processed)
+            else:
+                result.append(processed)
+
+    return result
+
+
+# --- Recursive Processor ---
+def process_element(element: str | list | object) -> str | list | object:
+    """Recursively processes an element, handling strings, text objects, lists, and dictionaries.
+
+    :param element: The element to process, which can be a string, list, or dictionary.
+    :return: The processed element, which may be a cleaned string, a list of paragraphs, or a dictionary.
+    """
+    if isinstance(element, str):
+        return process_string(element)
+    if isinstance(element, dict) and "text" in element:
```

Comment:
This needs some clarification or better names as it does process the dict twice (based on two conditions) in here, text_object is a dict too, the only difference is it has `text` key in it.